src/main/scala/is/hail/stats/BaldingNicholsModel.scala

```diff
@@ -78,35 +79,92 @@ object BaldingNicholsModel {
 
-    val rdd = sc.parallelize((0 until M).map(m => (m, Rand.randInt.draw())), nPartitions)
-      .map { case (m, perVariantSeed) =>
-        val perVariantRandomGenerator = new JDKRandomGenerator
-        perVariantRandomGenerator.setSeed(perVariantSeed)
-        val perVariantRandomBasis = new RandBasis(perVariantRandomGenerator)
+    val rowType = TStruct(
+      "pk" -> TLocus(gr),
+      "v" -> TVariant(gr),
+      "va" -> TStruct("ancestralAF" -> TFloat64(), "AF" -> TArray(TFloat64())),
+      "gs" -> !TArray(TStruct("GT" -> TCall())))
 
-        val ancestralAF = af_dist.getBreezeDist(perVariantRandomBasis).draw()
 
-        val popAF_k = (0 until K).map { k =>
-          new Beta(ancestralAF * Fst1_kBc.value(k), (1 - ancestralAF) * Fst1_kBc.value(k))(perVariantRandomBasis).draw()
+    val rdd = sc.parallelize((0 until M).map(m => (m, Rand.randInt.draw())), nPartitions)
+      .mapPartitions { it =>
+
+        val rvb = new RegionValueBuilder(MemoryBuffer())
+
+        it.map { case (m, perVariantSeed) =>
+          val perVariantRandomGenerator = new JDKRandomGenerator
+          perVariantRandomGenerator.setSeed(perVariantSeed)
+          val perVariantRandomBasis = new RandBasis(perVariantRandomGenerator)
+
+          val ancestralAF = af_dist.getBreezeDist(perVariantRandomBasis).draw()
+
+          val popAF_k = (0 until K).map { k =>
+            new Beta(ancestralAF * Fst1_kBc.value(k), (1 - ancestralAF) * Fst1_kBc.value(k))(perVariantRandomBasis).draw()
+          }
+
+          rvb.clear()
+          rvb.start(rowType)
+          rvb.startStruct()
+
+          // locus
+          rvb.startStruct()
+          rvb.addString("1")
+          rvb.addInt(m + 1)
+          rvb.endStruct()
+
+          // variant
+          rvb.startStruct()
+          rvb.addString("1")
+          rvb.addInt(m + 1)
+          rvb.addString("A")
+          rvb.startArray(1)
+          rvb.startStruct()
+          rvb.addString("A")
+          rvb.addString("C")
+          rvb.endStruct()
+          rvb.endArray()
+          rvb.endStruct()
+
+          // va
+          rvb.startStruct()
+          rvb.addDouble(ancestralAF)
+          rvb.startArray(popAF_k.length)
+          var i = 0
+          while (i < popAF_k.length) {
+            rvb.addDouble(popAF_k(i))
+            i += 1
+          }
+          rvb.endArray()
+          rvb.endStruct()
+
+          // gs
+          rvb.startArray(N)
+          i = 0
+          while (i < N) {
+            val p = popAF_k(popOfSample_nBc.value(i))
+            val pSq = p * p
+            val x = new Uniform(0, 1)(perVariantRandomBasis).draw()
+            val gt =
+              if (x < pSq)
+                2
+              else if (x > 2 * p - pSq)
+                0
+              else
+                1
+
+            rvb.startStruct()
+            rvb.addInt(gt)
+            rvb.endStruct()
+            i += 1
+          }
+          rvb.endArray()
+          rvb.endStruct()
+          rvb.result()
```

Comment:
We should delete this.  It conses.

Reply:
What's the preferred alternative? We could also allocate a region value once in the builder and mutate it.

Reply:
Yeah, allocate a region value in the mapPartitions, and then:

```
  rv.setOffset(rvb.end())
  rv
```

There's also `RegionValue.set(region: MemoryBuffer, offset: Long)`.